apps/emqx_s3/src/emqx_s3_client_http.erl

```diff
@@ -0,0 +1,245 @@
+                reason => Reason,
+                timeout => Timeout,
+                request => Method,
+                url => Url0,
+                time => tdiff(T0)
+            }),
+            Error
+    catch
+        error:Reason ->
+            ?SLOG(warning, #{
+                msg => "s3_http_request_fail",
+                reason => Reason,
+                timeout => Timeout,
+                method => Method,
+                url => Url0
+            }),
+            {error, Reason}
+    end.
+
+tdiff(T0) ->
+    T1 = erlang:monotonic_time(),
+    erlang:convert_time_unit(T1 - T0, native, millisecond).
+
+request_options(
+    Client = #{
+        pool := PoolName,
+        socket := SocketOpts
+    }
+) ->
+    [
+        {pool, PoolName},
+        {connect_timeout, connect_timeout(Client)},
+        {checkout_timeout, connect_timeout(Client)},
+        {recv_timeout, request_timeout(Client)},
+        {follow_redirect, false}
+    ] ++
+        [{connect_options, SocketOpts} || SocketOpts =/= []] ++
+        transport_options(Client).
+
+transport_options(#{transport := ssl, ssl := SSLOpts}) ->
+    [{ssl_options, SSLOpts}, {insecure, false}];
+transport_options(#{transport := tcp}) ->
+    [].
+
+%%
+
+connect_timeout(C) ->
+    maps:get(connect_timeout, C, ?DEFAULT_CONNECT_TIMEOUT).
+
+request_timeout(C) ->
+    maps:get(request_timeout, C, ?DEFAULT_REQUEST_TIMEOUT).
+
+%%
+
+start_pool(PoolName, ProfileConfig) ->
+    PoolConfig = pool_config(ProfileConfig),
+    ?SLOG(debug, #{msg => "s3_starting_http_pool", pool_name => PoolName, config => PoolConfig}),
+    case hackney_pool:start_pool(PoolName, PoolConfig) of
+        ok ->
+            ?SLOG(info, #{msg => "s3_start_http_pool_success", pool_name => PoolName}),
+            ok;
+        {error, _} = Error ->
+            ?SLOG(error, #{msg => "s3_start_http_pool_fail", pool_name => PoolName, error => Error}),
+            Error
+    end.
+
+update_pool(PoolName, ProfileConfig) ->
+    PoolConfig = pool_config(ProfileConfig),
+    ?SLOG(debug, #{msg => "s3_updating_http_pool", pool_name => PoolName, config => PoolConfig}),
+    lists:foldl(
+        fun
+            (PoolSetting, ok) -> update_pool_setting(PoolName, PoolSetting);
+            (_, Error) -> Error
+        end,
+        ok,
+        PoolConfig
+    ).
+
+update_pool_setting(PoolName, {max_connections, PoolSize}) ->
+    hackney_pool:set_max_connections(PoolName, PoolSize);
+update_pool_setting(PoolName, {timeout, Keepalive}) ->
+    hackney_pool:set_timeout(PoolName, Keepalive).
+
+stop_pool(PoolName) ->
+    case hackney_pool:stop_pool(PoolName) of
+        ok ->
+            ok;
+        {error, Reason} ->
+            ?SLOG(error, #{msg => "s3_stop_http_pool_fail", pool_name => PoolName, reason => Reason}),
+            ok
+    end.
+
+pool_config(#{
+    transport_options := #{
+        pool_type := _PoolType,
+        pool_size := PoolSize
+    }
+}) ->
+    [
```

Comment:
So hackney pools are not per-host, and we don't need all the "switching pool" mechanics?

Reply:
Right.

Reply:
That's a nice simplification